File: ios/SuiteBDE/UI/Components/Clubs/ClubDetailsView.swift
//
//  ClubDetailsView.swift
//  BDE
//
//  Created by Nathan Fallet on 12/02/2024.
//  Copyright © 2024 orgName. All rights reserved.
//

import SwiftUI
import GuimauveUI
import shared

struct ClubDetailsView: View {
    
    let club: CommonsClub
    let users: [CommonsUserInClub]
    let loadMore: (CoreUUID) -> Void
    let onJoinLeaveClicked: () -> Void
    
    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 16) {
                VStack(alignment: .leading) {
                    HStack {
                        Text("clubs_information")
                            .font(.title2)
                        Spacer()
                    }
                    Text((club.usersCount != 1 ? "clubs_members" : "clubs_member").localized().format(club.usersCount))
                        .foregroundStyle(.secondary)
                }
                Text(club.description_)
                    .multilineTextAlignment(.leading)
                
                Button(
                    club.isMember?.boolValue ?? false ? "clubs_button_leave" : "clubs_button_join",
                    action: onJoinLeaveClicked
                )
                .buttonStyle(DefaultButtonStyle(
                    filled: club.isMember?.boolValue != true
                ))
                
                HStack {
                    Text("clubs_information_members")
                        .font(.title2)
                    Spacer()
                }
                LazyVGrid(
                    columns: [GridItem(.adaptive(minimum: 300, maximum: 400))],
                    alignment: .leading,
                    spacing: 16
                ) {
                    ForEach(users, id: \.userId) { userInClub in
                        if let user = userInClub.user {
                            DefaultNavigationLink(destination: UserView(viewModel:
                                KoinApplication.shared.koin.userViewModel(associationId: user.associationId, userId: user.id)
                            )) {
                                UserCard(
                                    user: user,
                                    customDescription: userInClub.role.name
                                )
                                .onAppear {
                                    loadMore(userInClub.id)
                                }
                            }
                        }
                    }
                }
            }
            .padding()
        }
        .defaultNavigationTitle(club.name)
        .defaultNavigationBackButtonHidden(false)
        .defaultNavigationImage {
            AsyncImage(
                url: URL(string: club.logo ?? ""),
                content: { image in
                    image
                        .resizable()
                        .aspectRatio(contentMode: .fill)
                },
                placeholder: {
                    Color(UIColor.systemGray3)
                        .overlay {
                            Text(club.name.initials)
                                .font(.title)
                        }
                }
            )
        }
    }
    
}

#Preview {
    DefaultNavigationView {
        ClubDetailsView(
            club: CommonsClub(
                id: CoreUUID(),
                associationId: CoreUUID(),
                name: "Club running",
                description: "Club running de l'ENSISA ! RDV tous les jeudis à la barrière du parking Werner",
                logo: "https://bdensisa.org/clubs/rev4fkzzd79u7glwk0l1agdoovm3s7yo/uploads/logo%20club%20run.jpeg",
                createdAt: Date().asKotlinxInstant,
                validated: true,
                usersCount: 12,
                isMember: true
            ),
            users: [
                CommonsUserInClub(
                    userId: CoreUUID(),
                    clubId: CoreUUID(),
                    roleId: CoreUUID(),
                    user: CommonsUser(
                        id: CoreUUID(),
                        associationId: CoreUUID(),
                        email: "",
                        password: nil,
                        firstName: "Nathan",
                        lastName: "Fallet",
                        superuser: true,
                        lastLoginAt: Date().asKotlinxInstant
                    ),
                    club: nil,
                    role: CommonsRoleInClub(
                        id: CoreUUID(),
                        clubId: CoreUUID(),
                        name: "Admin",
                        admin: true,
                        default: false
                    )
                ),
                CommonsUserInClub(
                    userId: CoreUUID(),
                    clubId: CoreUUID(),
                    roleId: CoreUUID(),
                    user: CommonsUser(
                        id: CoreUUID(),
                        associationId: CoreUUID(),
                        email: "",
                        password: nil,
                        firstName: "Maxime",
                        lastName: "Sanciaume",
                        superuser: true,
                        lastLoginAt: Date().asKotlinxInstant
                    ),
                    club: nil,
                    role: CommonsRoleInClub(
                        id: CoreUUID(),
                        clubId: CoreUUID(),
                        name: "Membre",
                        admin: false,
                        default: true
                    )
                )
            ],
            loadMore: { _ in },
            onJoinLeaveClicked: {}
        )
    }
}

File: android/src/main/kotlin/com/suitebde/ui/components/clubs/ClubDetailsView.kt
package com.suitebde.ui.components.clubs

import androidx.compose.foundation.ExperimentalFoundationApi
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import coil.compose.AsyncImage
import com.suitebde.R
import com.suitebde.models.clubs.Club
import com.suitebde.models.clubs.UserInClub
import com.suitebde.ui.components.navigation.DefaultNavigationBar
import com.suitebde.ui.components.users.UserCard
import dev.kaccelero.models.UUID
import kotlinx.datetime.Clock

@OptIn(ExperimentalFoundationApi::class, ExperimentalMaterial3Api::class)
@Composable
@Suppress("FunctionName")
fun ClubDetailsView(
    club: Club,
    users: List<UserInClub>,
    loadMore: (UUID) -> Unit,
    onJoinLeaveClicked: () -> Unit,
    navigateUp: () -> Unit,
    modifier: Modifier = Modifier,
) {

    LazyColumn(
        verticalArrangement = Arrangement.spacedBy(16.dp),
        modifier = modifier
    ) {
        stickyHeader {
            DefaultNavigationBar(
                title = club.name,
                navigateUp = navigateUp,
                image = {
                    AsyncImage(
                        model = club.logo ?: "",
                        placeholder = painterResource(R.drawable.default_event_image),
                        error = painterResource(R.drawable.default_event_image),
                        contentDescription = null,
                        contentScale = ContentScale.Crop,
                        modifier = it
                    )
                }
            )
        }
        item {
            Column(
                modifier = Modifier.padding(horizontal = 16.dp)
            ) {
                Text(
                    text = stringResource(R.string.clubs_information),
                    style = MaterialTheme.typography.titleMedium
                )
                Text(
                    text = stringResource(if (club.usersCount != 1L) R.string.clubs_members else R.string.clubs_member)
                        .format(club.usersCount),
                    color = Color.Gray
                )
            }
        }
        item {
            Text(
                text = club.description,
                modifier = Modifier.padding(horizontal = 16.dp)
            )
        }
        item {
            if (club.isMember == true) OutlinedButton(
                onClick = onJoinLeaveClicked,
                modifier = Modifier
                    .padding(horizontal = 16.dp)
                    .fillMaxWidth(),
                shape = MaterialTheme.shapes.medium,
                content = { Text(stringResource(R.string.clubs_button_leave)) }
            ) else Button(
                onClick = onJoinLeaveClicked,
                modifier = Modifier
                    .padding(horizontal = 16.dp)
                    .fillMaxWidth(),
                shape = MaterialTheme.shapes.medium,
                content = { Text(stringResource(R.string.clubs_button_join)) }
            )
        }
        item {
            Text(
                text = stringResource(R.string.clubs_information_members),
                style = MaterialTheme.typography.titleMedium,
                modifier = Modifier.padding(horizontal = 16.dp)
            )
        }
        items(users) {
            it.user?.let { user ->
                UserCard(
                    user = user,
                    customDescription = it.role.name,
                    modifier = Modifier.padding(horizontal = 16.dp)
                )
                loadMore(it.id)
            }
        }

        item {
            Spacer(modifier = Modifier)
        }
    }

}

@Preview
@Composable
@Suppress("FunctionName")
fun ClubDetailsViewPreview() {
    ClubDetailsView(
        club = Club(
            id = UUID(),
            associationId = UUID(),
            name = "Club running",
            description = "A cool club",
            logo = "https://bdensisa.org/clubs/rev4fkzzd79u7glwk0l1agdoovm3s7yo/uploads/logo%20club%20run.jpeg",
            createdAt = Clock.System.now(),
            validated = true,
            usersCount = 12,
            isMember = true
        ),
        users = listOf(),
        loadMore = {},
        onJoinLeaveClicked = {},
        navigateUp = {}
    )
}
